angelsrefining/prototypes/angels-functions.lua

```diff
@@ -540,10 +639,12 @@ function angelsmods.functions.create_liquid_recipe_icon(bot_molecules_icon, tint
   -- allows a string of max 3 characters for default tints
   if tints then
     if type(tints) ~= "table" then
+      local reference = get_molecule_codes(tints)
+      ref = {reference[1].form,reference[2].form,reference[3].form} --don't go past the 3
       tints = {
-        top = unify_tint((icon_tint_table[string.sub(tints, 1, 1)] or {})[1]),
-        mid = unify_tint((icon_tint_table[string.sub(tints, 2, 2)] or {})[2]),
-        bot = unify_tint((icon_tint_table[string.sub(tints, 3, 3)] or {})[3])
+        top = unify_tint(icon_tints_table[ref[1]][1] or {}),
+        mid = unify_tint(icon_tints_table[ref[2]][2] or {}),
+        bot = unify_tint(icon_tints_table[ref[3]][3] or {})
       }

```

Comment:
The ref step is obsolete, it also does not provide a fallback if there are less than 3 refs defined?
```suggestion
      local reference = get_molecule_codes(tints)
      tints = {
        top = unify_tint(icon_tints_table[(reference[1] or {form="unknown"}).form][1] or {}),
        mid = unify_tint(icon_tints_table[(reference[2] or {form="unknown"}).form][2] or {}),
        bot = unify_tint(icon_tints_table[(reference[3] or {form="unknown"}).form][3] or {})
      }
```

Reply:
im a bit confused by the {form="unknown"} bit, im guessing it is to either throw an error, or do you want a default value?

Reply:
- If you only provide 1 color reference, then `reference[2]` equals `nil`.
- The next step is to index it: `reference[2].form`, which would throw an error 'attempt to index a nil value', thus give it a default table to resolve this error: `(reference[2] or {}).form`
- Now you use this result to index the `icon_tints_table`: `icon_tints_table[(reference[2] or {}).form]`, which again could yield an error 'attempt to index icon_tints_table with .form, a nil value'. This can also be resolved to give it again a default value, I opted here for 'unknown' as it is fairly sure to have no value assigned to that key: `icon_tints_table`: `icon_tints_table[(reference[2] or {}).form or "unknown"]`.
- Now I simplify that last statement as every 'or' is an if statement creating a branch: `icon_tints_table[(reference[2] or {form="unknown"}).form]`. This still handles all edge cases, just compressed all 'fallback' values into one or statement.